### PR TITLE
feat: config reloading

### PR DESCRIPTION
**Pull request by cstockton**
## What kind of change does this PR introduce?

File based configuration reloading using fsnotify. 

## What is the current behavior?

Currently the Auth config is loaded once from the environment or file (-c flag) and persists until the service is restarted.

## What is the new behavior?

A new optional flag (long: `--watch-dir`, short: `-w`) has been added. When present any files with a ".env" suffix will be loaded into the environment before the `*GlobalConfiguration` is created, otherwise existing behavior is preserved.

In addition when the watch-dir flag is present a goroutine will be started in serve_cmd.go and begin blocking on a call to `(*Reloader).Watch` with a callback function that accepts a `*conf.GlobalConfiguration object`. Each time this function is called we create a new `*api.API` object and store it within our `AtomicHandler`, previously given as the root handler to the `*http.Server`.

The Reloader uses some simple heuristics to deal with a few edge cases, an overview:

  - At most 1 configuration reload may happen per 10 seconds with a +-1s margin of error.
  - After a file within `--watch-dir` has changed the 10 second grace period begins. After that it will reload the config.
  - Config reloads first sort each file by name then processes them in sequence.
  - Directories within `--watch-dir` are ignored during config reloading.
    - Implementation quirk: directory changes can trigger a config reload, as I don't stat fsnotify events. This and similar superfulous reloads could be easily fixed by storing a snapshot of `os.Environ()` after successful reloads to compare with the latest via `slices.Equal()` before reloading.
  - Files that do not end with a `.env` suffix are ignored.
  - It handles the removal or renaming of the `-watch-dir` during runtime, but an error message will be printed every 10 seconds as long as it's missing.
  - The config file passed with -c is only loaded once. Live reloads only read the config dir. Meaning it would be possible to create a config dir change that results in a new final configuration on the next reload due to the persistence of `os.Environ()`.

### CODE REVIEW

**Comment by kangmingtay** [comment]
@cstockton seems like some of the gosec tests are broken

**Review by hf — APPROVED** [review]

File: cmd/root_cmd.go [hunk 1]
```diff
@@ -22,8 +25,8 @@ var rootCmd = cobra.Command{
 // RootCommand will setup and return the root command
 func RootCommand() *cobra.Command {
 	rootCmd.AddCommand(&serveCmd, &migrateCmd, &versionCmd, adminCmd())
-	rootCmd.PersistentFlags().StringVarP(&configFile, "config", "c", "", "the config file to use")
-
+	rootCmd.PersistentFlags().StringVarP(&configFile, "config", "c", "", "base configuration file to load")
+	rootCmd.PersistentFlags().StringVarP(&watchDir, "watch-dir", "w", "", "directory containing a sorted list of config files to watch for changes")
```
Comment (hf) [inline cmd/root_cmd.go:29]: Last minute change... maybe it's best to call this `config-dir` instead?

File: internal/reloader/testdata/50_example.env [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+GOTRUE_HOOK_CUSTOM_SMS_PROVIDER_ENABLED=false
+GOTRUE_HOOK_CUSTOM_SMS_PROVIDER_URI=""
+# Only for HTTPS Hooks
+GOTRUE_HOOK_CUSTOM_SMS_PROVIDER_SECRET=""
+
+
+# Test OTP Config
+GOTRUE_SMS_TEST_OTP="<phone-1>:<otp-1>, <phone-2>:<otp-2>..."
+GOTRUE_SMS_TEST_OTP_VALID_UNTIL="2050-01-01T01:00:00Z" # (e.g. 2023-09-29T08:14:06Z)
```
Comment (hf) [inline internal/reloader/testdata/50_example.env:235]:
```suggestion
GOTRUE_SMS_TEST_OTP_VALID_UNTIL="2050-01-01T01:00:00Z" # (e.g. 2023-09-29T08:14:06Z)

```

File: cmd/root_cmd.go (hunk 1 above)
Comment (cstockton) [reply cmd/root_cmd.go:29]: Sure that's an easy change.

**Review by kangmingtay — APPROVED** [review]
nice one @cstockton !

**Pull request merged**